File: ROS2/CHAPT5/src/demo_cpp_tf/src/dynamic_tf_boardcaster.cpp
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "rclcpp/rclcpp.hpp" //提供消息接口
#include "tf2/LinearMath/Quaternion.h" //提供tf2::Quaternion类
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp" //消息接口转换函数
#include "tf2_ros/transform_broadcaster.h" //坐标广播器类
#include "chrono"

using namespace std::chrono_literals;

class TFBoardcaster:public rclcpp::Node{
    private:
        std::shared_ptr<tf2_ros::TransformBroadcaster>boardcaster;
        rclcpp::TimerBase::SharedPtr timer;

    public:
        TFBoardcaster():Node("tf_boardcaster")
        {
            this->boardcaster=std::make_shared<tf2_ros::TransformBroadcaster>(this);
            // this->publish_tf();
            timer=this->create_wall_timer(100ms,std::bind(&TFBoardcaster::publish_tf,this));
        }

        void publish_tf()
        {
            geometry_msgs::msg::TransformStamped transform;
            transform.header.stamp=this->get_clock()->now();
            transform.header.frame_id="map";
            transform.child_frame_id="base_link";
            transform.transform.translation.x=2.0;
            transform.transform.translation.y=3.0;
            transform.transform.translation.z=0.0;           
            tf2::Quaternion q;
            q.setRPY(0.0,0.0,30*M_PI/180.0);
            transform.transform.rotation=tf2::toMsg(q);
            this->boardcaster->sendTransform(transform);
        }
};


int main(int argc,char* argv[])
{
    rclcpp::init(argc,argv);
    auto node =std::make_shared<TFBoardcaster>();
    rclcpp::spin(node);
    rclcpp::shutdown();
}

File: ROS2/CHAPT5/src/demo_python_tf/demo_python_tf/dynamic_tf_boardcaster.py
import rclpy
from rclpy.node import Node
from tf2_ros import TransformBroadcaster # 静态坐标发布器
from geometry_msgs.msg import TransformStamped # 消息接口
from tf_transformations import quaternion_from_euler # 欧拉角转四元函数
import math # 角度转弧度函数

class TFBoardcaster(Node):
    def __init__(self):
        super().__init__("tf_boradcaster")
        self.boardcaster=TransformBroadcaster(self)
        # self.publish_static_tf()
        self.timer=self.create_timer(0.01,self.publish_tf)

    
    def publish_tf(self):
        """
        发布静态TF 从camera_link到bottle_link之间的坐标关系
        """
        transform=TransformStamped()
        transform.header.frame_id="camera_link"
        transform.child_frame_id="bottle_link"
        transform.header.stamp=self.get_clock().now().to_msg()

        transform.transform.translation.x=0.2
        transform.transform.translation.y=0.3
        transform.transform.translation.z=0.5        
        # 欧拉角转四元数 x,y,z,w
        q=quaternion_from_euler(0,0,0)
        # 旋转部分进行赋值

        transform.transform.rotation.x=q[0]
        transform.transform.rotation.y=q[1]
        transform.transform.rotation.z=q[2]
        transform.transform.rotation.w=q[3]

        # 静态坐标关系发布出去
        self.boardcaster.sendTransform(transform)
        self.get_logger().info(f"发布TF:{transform}")

def main():
    rclpy.init()
    node=TFBoardcaster()
    rclpy.spin(node)
    rclpy.shutdown()